session-replay/Sources/DatadogSessionReplay/DatadogCoreIntegration/RUMContextReceiver.swift

```diff
@@ -0,0 +1,96 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+@testable import Datadog
+
+/// The RUM context received from `DatadogCore`.
+internal struct RUMContext: Equatable {
+    /// Current RUM application ID - standard UUID string, lowecased.
+    let applicationID: String
+    /// Current RUM session ID - standard UUID string, lowecased.
+    let sessionID: String
+    /// Current RUM view ID - standard UUID string, lowecased.
+    let viewID: String
+}
+
+/// An observer notifying on`RUMContext` changes.
+internal protocol RUMContextObserver {
+    /// Starts notifying on distinct changes to `RUMContext`.
+    ///
+    /// - Parameters:
+    ///   - queue: a queue to call `notify` block on
+    ///   - notify: a closure receiving new `RUMContext` or `nil` if current RUM session is not sampled
+    func observe(on queue: Queue, notify: @escaping (RUMContext?) -> Void)
+}
+
+/// Receives RUM context from `DatadogCore` and notifies it through `RUMContextObserver` interface.
+internal class RUMContextReceiver: FeatureMessageReceiver, RUMContextObserver {
+    /// Notifies new `RUMContext` or `nil` if current RUM session is not sampled.
+    private var onNew: ((RUMContext?) -> Void)?
+    private var previous: RUMContext?
+
+    // MARK: - FeatureMessageReceiver
+
+    func receive(message: FeatureMessage, from core: DatadogCoreProtocol) -> Bool {
+        guard let rumBaggage = message.contextMessage?.rumBaggage else {
+            // No RUM baggage in the message
+            return false
+        }
+
+        // Extract the `RUMContext` or `nil` if RUM session is not sampled:
+        let new = rumBaggage.rumContext
+
+        // Notify only if it has changed:
+        if new != previous {
```

Comment:
I think this should be handle by core: send a message only if the feature attributes changed. Not a change request, I will add a ticket for enhancing `FeatureBaggage`:

1. Enable serialisation/deserialisation of `Codable` in `FeatureBaggage`. It will be **very (very)** convenient for Features to have a validation mechanism such as `Codable` out-of-the-box. It will prevent from having to repeat properties checks in Features.
2. Make `FeatureBaggage` compliant to `Equitable` and on send a message if the baggage changes.

This will greatly simplify your message processing!

Reply:
Could be - having this done from Core will work for SR as well 👍 and yeah it could be a way to avoid this (☝️) code duplication in other products. As it's not a blocker I will keep it like this for now, as for the "uploader" topic I will **perhaps** need to change this logic anyway (receive entire context instead of just the RUM part).